cmd/vulcan-snyk/main.go

```diff
@@ -0,0 +1,261 @@
+
+	check "github.com/adevinta/vulcan-check-sdk"
+	"github.com/adevinta/vulcan-check-sdk/state"
+	report "github.com/adevinta/vulcan-report"
+
+	"github.com/adevinta/vulcan-check-sdk/helpers/command"
+)
+
+var (
+	checkName = "vulcan-snyk"
+	logger    = check.NewCheckLog(checkName)
+)
+
+func main() {
+	run := func(ctx context.Context, target string, optJSON string, state state.State) (err error) {
+		logger.WithFields(logrus.Fields{
+			"repository": target,
+		}).Debug("testing repository")
+
+		if target == "" {
+			return errors.New("check target missing")
+		}
+
+		targetURL, err := url.Parse(target)
+		if err != nil {
+			return err
+		}
+
+		var auth *http.BasicAuth
+		if targetURL.Host == "github.mpi-internal.com" {
+			auth = &http.BasicAuth{
+				Username: "username", // Can be anything except blank.
+				Password: os.Getenv("GITHUB_ENTERPRISE_TOKEN"),
+			}
+		}
+
+		repoPath := filepath.Join("/tmp", filepath.Base(targetURL.Path))
+		err = os.RemoveAll(repoPath)
+		if err != nil {
+			return err
+		}
+
+		if err := os.Mkdir(repoPath, 0755); err != nil {
+			return err
+		}
+
+		_, err = git.PlainClone(repoPath, false, &git.CloneOptions{
+			URL:  target,
+			Auth: auth,
+		})
+		if err != nil {
+			return err
+		}
+
+		if os.ExpandEnv("$SNYK_TOKEN") == "" {
+			return fmt.Errorf("SNYK_TOKEN is not set")
+		}
+
+		output, code, err := command.Execute(ctx, logger, "snyk", append([]string{"auth", os.ExpandEnv("$SNYK_TOKEN")})...)
+		if code != 0 {
+			return fmt.Errorf("%s", output)
+		}
+
+		logger.Infof("auth: %s", output)
+		if err != nil {
+			return err
+		}
+
+		output, _, _ = command.Execute(ctx, logger, "snyk", append([]string{"test", repoPath, "--all-sub-projects", "--json"})...)
+
+		r := SnykResponse{}
+		err = json.Unmarshal(output, &r)
+		if err != nil {
+			return err
+		}
+
+		// Group vulnerabilities by their Snyk Reference ID and module name, also filter out license issues
+		snykVulnerabilitiesMap := make(map[string]map[string][]SnykVulnerability)
+		for _, v := range r.Vulnerabilities {
+			if v.Type == "license" {
+				continue
+			}
+
+			_, ok := snykVulnerabilitiesMap[v.ID]
+			if !ok {
+				snykVulnerabilitiesMap[v.ID] = make(map[string][]SnykVulnerability)
+			}
+
+			snykVulnerabilitiesMap[v.ID][v.ModuleName] = append(snykVulnerabilitiesMap[v.ID][v.ModuleName], v)
+		}
+
+		vulns := []report.Vulnerability{}
+		// for each pair of (snyk vuln & module name), create a vulcan vuln
+		for _, snykModulesMap := range snykVulnerabilitiesMap {
+			for moduleName, snykIssues := range snykModulesMap {
+				vulcanVulnerability := &report.Vulnerability{}
+
+				vulcanVulnerability.Summary = snykIssues[0].Title + ": " + moduleName
+				vulcanVulnerability.Description = extractOverview([]byte(snykIssues[0].Description))
+				vulcanVulnerability.Details = extractDetails([]byte(snykIssues[0].Description))
+				vulcanVulnerability.ImpactDetails = createImpactDetails(snykIssues)
```

Comment:
I don't understand why you iterate through all of the issues here? Won't that yield the full list of dependencies that introduced issues instead of the dependencies that introduced that particular issue? Anyhow, I think impact details is meant to explain the potential impact of the vulnerability, I think this should go in resources or maybe details.

Reply:
Hi @gguillemas, Snyk will report one issue for each vulnerable library found in the dependency tree.
For example, imagine that `erubis` is imported by 9 libraries used by `Project Foo`, so Snyk will return 9 entries for that. Those entries will contain repeated information about the issue itself: description, remmediation, etc and also a `from` field, which contains specific information about which dependencies introduced that issue. 

Also, even if Snyk return multiple JSON entries for the same issue through the CLI, they are rendering a single Issue (with an aggregated `Introduced paths` field) in their UI report.

Regarding wether this information fits the `impact details` field, you are right. I have mixed `details` and `impact_details`. They are inverted. I will fix that.

Reply:
Cool! I understand :+1:

In that case, I would really use the resources table, as each library is a resource that is affected by the vulnerability. I think a great example is the AWS Trusted Advisor check.